system/lib/wasmfs/syscalls.cpp

```diff
@@ -513,4 +536,73 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_getdents64(long fd, long dirp, long count) {
+  auto openFile = wasmFS.getLockedFileTable()[fd];
+
+  if (!openFile) {
+    return -EBADF;
+  }
+  dirent* result = (dirent*)dirp;
+
+  // Check if the result buffer is too small.
+  if (count / sizeof(dirent) == 0) {
+    return -EINVAL;
+  }
+
+  auto file = openFile.locked().getFile();
+
+  auto directory = file->dynCast<Directory>();
+
+  if (!directory) {
+    return -ENOTDIR;
+  }
+
+  // Hold the locked directory for the duration of this operation.
+  auto lockedDir = directory->locked();
+
+  off_t bytesRead = 0;
+  // A directory's position corresponds to the index in its entries vector.
+  int index = openFile.locked().position();
+
+  // In the root directory, ".." refers to itself.
+  auto dotdot =
+    file == wasmFS.getRootDirectory() ? file : lockedDir.getParent();
+
+  // If the directory is unlinked then the parent pointer should be null.
+  if (!dotdot) {
+    return -ENOENT;
+  }
+
+  // There are always two hardcoded directories "." and ".."
+  std::vector<Directory::entry> entries = {{".", file}, {"..", dotdot}};
+  auto dirEntries = lockedDir.getEntries();
+  entries.insert(entries.end(), dirEntries.begin(), dirEntries.end());
+
+#ifdef WASMFS_DEBUG
+  for (auto pair : entries) {
+    emscripten_console_log(pair.name.c_str());
+  }
+#endif
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    // Set inode number to the file pointer. This gives a unique inode number
+    // for each file in the simplest possible way.
```

Comment:
```suggestion
```
This comment is now stale. It is not needed here, and it just adds risk of becoming incorrect some day if we change the impl in the other file.

Reply:
Moved it to the `getIno` function. I think it's relevant there.